mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp

```diff
@@ -1890,9 +1890,67 @@ struct FoldTransposeWithTranspose : OpRewritePattern<linalg::TransposeOp> {
   }
 };
 
+/// This pattern reduces the cost of transpose by swapping the order of
+/// broadcast and transpose:
+///   transpose(broadcast(input)) -> broadcast(transpose(input))
```

Comment:
I wouldn't say "reduces the cost of transpose" as this may turn a transpose with a performant implementation into a simpler one that may or may not have a performant implementation. I would just describe this as a canonicalization pattern.